faidx.c

```diff
@@ -360,55 +435,105 @@ int faidx_core(int argc, char *argv[], enum fai_format_options format)
         else
             fprintf(stderr, "\n");
 
-        return EXIT_FAILURE;
+        goto exit2;
     }
 
     /** output file provided by user */
     if( output_file != NULL ) {
         if( strcmp( output_file, argv[optind] ) == 0 ) {
             fprintf(stderr,"[faidx] Same input/output : %s\n", output_file);
-            return EXIT_FAILURE;
+            goto exit2;
+        }
+        if (!out.isbgzip) {
+            out.fp = fopen( output_file, "w" );
+        }
+        else {
+            hts_opt *opts = (hts_opt *)(out.gopt->out.specific);
+            char mode[13] = "w";
+            int level = 0;
+            while (opts) {
+                if (opts->opt == HTS_OPT_COMPRESSION_LEVEL) {   //compression level
+                    level = opts->val.i;
+                    break;
+                }
+                opts = opts->next;
+            }
+            if (level > 0) {
+                snprintf(mode, sizeof(mode), "w%d", level);     //pass compression with mode
+            }
+            out.bgzf_fp = bgzf_open(output_file, mode);
         }
 
-        file_out = fopen( output_file, "w" );
-
-        if( file_out == NULL) {
+        if( (!out.isbgzip && out.fp == NULL) || (out.isbgzip && out.bgzf_fp == NULL)) {
             fprintf(stderr,"[faidx] Cannot open \"%s\" for writing :%s.\n", output_file, strerror(errno) );
-            return EXIT_FAILURE;
+            goto exit2;
+        }
+        //set threading if set
+        if (out.isbgzip && out.gopt->nthreads > 0) {
+            if (bgzf_mt(out.bgzf_fp, out.gopt->nthreads, 256/*unused*/) < 0) {
+                fprintf(stderr,"[faidx] Failed to set threads\n");
+                goto exit1;
+            }
```

Comment:
This made me recall that faidx does support reading bgzf files too, but it's never had the option to multi-thread the decoder.

However `perf record` implies libdeflate takes about 30% of the CPU time up, so I doubt it's really relevant.  If we're also compressing, then it'll be an even lower percentage of the CPU so attempting to share a thread pool probably isn't particularly useful.

So after musing a bit, I think this is probably the best way to handle this already.

Reply:
updated to use a thread pool in reading as well as writing compressed files.